### DISCUSSION TITLE
convergence (polynomial vs log function) and a question about KKS model physics

### DISCUSSION THREAD

**Discussion by bosxered** [idaholab/moose#24441] [Q&A Modules: Phase field] [answered]
Hello Moose Experts,

I tried to study KKS_system with a test example file,
/modules/phase_field/test/tests/KKS_system/kks_example_split.i

I have three questions.

**1. Why does not the log bulk free energy function converge?**
-  I changed the free energy function in the input file, kks_example_split.i just like below.

```
[Materials]
  # Free energy of the matrix
  [./fm]
    type = DerivativeParsedMaterial
    property_name = fm
    coupled_variables = 'cm'
    expression = '0.4009*(1.9956*cm+cm*log(cm)+(1-cm)*log(1-cm))+0.0441'
    outputs = exodus
  [../]

  # Free energy of the delta phase
  [./fd]
    type = DerivativeParsedMaterial
    property_name = fd
    coupled_variables = 'cd'
#    expression = '(0.9-cd)^2'
    expression = '0.4009*(-1.9956*cd+cd*log(cd)+(1-cd)*log(1-cd))+0.8441'
    outputs = exodus
  [../]
```
- Before changes, they were polynomial functions like  '(0.9-cd)^2'.
- I changed them because, in many papers (even the first proposal of KKS phase field model(S.Kim et al., PRE 60, 6, 1999)), log free energy function was used.
- I don't know why this does not converge like below.

```
Time Step 1, time = 2.32831e-11, dt = 2.32831e-11
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 1.16415e-11, dt = 1.16415e-11
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 5.82077e-12, dt = 5.82077e-12
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 2.91038e-12, dt = 2.91038e-12
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 1.45519e-12, dt = 1.45519e-12
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 7.27596e-13, dt = 7.27596e-13
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 3.63798e-13, dt = 3.63798e-13
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 1.81899e-13, dt = 1.81899e-13
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge

Time Step 1, time = 1e-13, dt = 1e-13
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
^[[31m Solve Did NOT Converge!^[[39m
Aborting as solve did not converge
```

For your information, I changed the polynomial function to the log function like below.
red +: original polynomial function
blue line: new log function (a1*(a2*cm+cm*log(cm)+(1-cm)*log(1-cm))+a3), where a1, a2, and a3 are constants.
![image](https://github.com/idaholab/moose/assets/79071957/59bb5e98-c095-4cab-9459-a8c398362fd0)
![image](https://github.com/idaholab/moose/assets/79071957/4e6fe5d1-4a70-4db5-b5b5-9579118162dc)

**2. What makes the phase transition?** 
Three pictures below are the simulation results of the input file, kks_example_split.i at 0.1s, 20s, and 40s.
![image](https://github.com/idaholab/moose/assets/79071957/c8ebe03a-cb13-48d5-8441-746b4338a4ca)

![image](https://github.com/idaholab/moose/assets/79071957/589da95d-3c8b-455c-bd6e-3fdfda43d8bb)

![image](https://github.com/idaholab/moose/assets/79071957/1f938053-99e0-4837-ad3f-7159e7a21f5b)

**If I want the yellow region to grow all over the system and maintain, what should the Material property be?
For example, barrier height(w) is high and ...etc.**

Also, It would be a great help if you let me know How I can make a rough guess before the simulation.

Thanks for reading this long discussion.

Best regard,
Jeonghwan.

**Comment by GiudGiud** [comment, marked as answer]
Hello

For your first question, your expression will create NaNs if cm <= 0 or cm >= 1. Same for cd
0.4009*(1.9956*cm+cm*log(cm)+(1-cm)*log(1-cm))+0.0441

So you need to bound cm and cd to make sure that does not happen.
We have created machinery in MOOSE to handle this.
In this material you will want to use tol_names and tol_values parameters to do this

Guillaume

**Reply by bosxered** [reply]
Thanks for your help!
